### PR TITLE
SNOW-821772: Add support for functions.flatten

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-821772

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Add `array_flatten` to `snowflake.snowpark.functions`

### CODE REVIEW

File: src/snowflake/snowpark/functions.py [hunk 1]
```diff
@@ -3356,6 +3356,19 @@ def array_max(array: ColumnOrName) -> Column:
     return builtin("array_max")(array)
 
 
+def array_flatten(array: ColumnOrName) -> Column:
+    """Returns a single array from an array or arrays. If the array is nested more than
+    two levels deep, then only a single level of nesting is removed.
+
+    Must enable parameter `ENABLE_ARRAY_FLATTEN_FUNCTION` in your session.
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/functions.py:3363]: Is this a public parameter that user can change?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/functions.py:3363]: It is. We are waiting for [BCR](https://snowflakecomputing.atlassian.net/browse/BCR-1239) to go through so setting this param will no longer be necessary.

File: tests/integ/test_function.py [hunk 2]
```diff
@@ -1109,6 +1110,21 @@ def test_array_min_max_functions(session):
     assert res == [Row(MAX_A="null")]
 
 
+@pytest.mark.xfail(reason="SNOW-844750 Waiting for BCR to complete", strict=False)
+def test_array_flatten(session):
+    df = session.create_dataframe([[[[1, 2, 3], [None], [4, 5]]],], schema=["a"])
+    df = df.select(array_flatten(df.a).as_("flatten_a"))
+    Utils.check_answer(df, [Row(FLATTEN_A='[\n  1,\n  2,\n  3,\n  null,\n  4,\n  5\n]')], statement_params={"ENABLE_ARRAY_FLATTEN_FUNCTION": True})
```
Comment (sfc-gh-sfan) [inline tests/integ/test_function.py:1117]: It might be better to directly enable this param on our test accounts.

**Review by sfc-gh-sfan — APPROVED** [review]
Let's make sure we have a ticket tracking the removal of this param.

**Pull request merged**